Fix `nn.Module.state_dict()`

lkct opened the pull request:
Fixes #72778

TODO
- [x] Waiting for a conclusion from discussion in the issue.
- [x] Still bugs in handling misplaced args. Need a re-design to cover all corner cases.

TODO changes
- [x] Put deprecated signature to the second.
- [x] Change to kwargs, positional deprecated
- [x] `DeprecationWarning` add comment on why not use it
- [x] Remove unnecessary comments.

albanD reviewed (changes requested):
Let's chat on the issue on how to resolve this first.
We can revisit the PR later if this solution is the way to go.

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1309,19 +1372,43 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
+        #   Actual values are:
+        # destination=pref, prefix=keep, keep_vars=False
+
+        if isinstance(destination, torch._six.string_classes):
```
Not sure what you're trying to do here.
The two arguments will have to be passed as kwargs in the final version of the function.
You can mark this with kwarg-only arguments: `def state_dict(self, *, prefix: str = ..., keep_vars: bool = ...)`.

lkct replied:
I thought it was the safest way to fix it by keeping both the original signatures valid.
Moving directly to kw-only can be an easier fix and I agree it will be the final version. However, it may break BC, e.g. some existing code could be using `mod.state_dict(some_custom_dict, 'foo.', True)`.
I need further confirmation on what signature should be exposed to users in this fix.

lkct replied:
Existing codes using `(prefix, keep_vars)` have no BC issues, because it would cause a crash if not using kwargs.

Existing code using `(destination, prefix, keep_vars)` can have positional args, so forcing kw-args may cause problems. However, this is the use case that we want to deprecate.

lkct replied:
A possible solution:
```python
def state_dict(destination, _prefix, _keep_vars, *, prefix, keep_vars):
    if prefix is provided: use prefix
    elif _prefix is provided: use prefix=_prefix
    else: use prefix=default
    # same for keep_vars
    if any destination, _prefix, _keep_vars: warning
```

albanD replied:
> I thought it was the safest way to fix it by keeping both the original signatures valid.

I think that unfortunately, the current overload annotations are wrong. And in particular, the second one is not valid. So that is fine to not support that (since it never worked).

For the fix, what about only warning today if "destination" is passed? Since the only way not to pass in destination is to pass the two others as kwargs. Once we delete destination, we can make the two others kwargs only.

So I think what we want to do here is what you have for the current signature
```python
def state_dict(*, prefix, keep_vars): ...
def state_dict(destination, prefix, keep_vars): ...  # This one is deprecated!
```

And the implementation can simply be
```python
def state_dict(self, *args, prefix='', keep_vars=False):
    if len(args)>0:
        # warn here that positional is deprecated
        destination = args[0]
        if len(args)>1:
            # check that prefix == ''
            prefix = args[1]
        # same for keeps var
        
```

What do you think?

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1278,8 +1278,47 @@ def _save_to_state_dict(self, destination, prefix, keep_vars):
+    # Keep only the second signature after removal
+    # The first overload is only for BC
+
+    # Comment on the comment below: it originated from the type stub in .pyi.
```
The comments below are not really needed.

lkct replied:
will fix

lkct commented on torch/nn/modules/module.py:
```diff
@@ -1309,19 +1372,43 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
+            destination = None
+
+        if destination is not None:
+            warnings.warn(
```
There are cases like this:
https://github.com/pytorch/pytorch/blob/87975d895c3f52cb33f48757fabefbb3d3f276f2/torch/nn/modules/container.py#L20-L22
However, it's not consistent throughout the pytorch repo. Use of `DeprecationWarning` also exists.
Should we force the warning to show (as it is now) or use `DeprecationWarning` that will only show on user config?
Maybe I should add the comment if it's decided that we're keeping the current.

albanD replied:
It's not super consistent indeed. I agree that we want this one to be loud as we're removing it. You can leave it as-is.

lkct commented on torch/nn/modules/module.py:
```diff
@@ -1309,19 +1372,43 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
+                "Argument destination is deprecated. nn.Module.state_dict will not accept it in the future. "
+                "Refer to https://pytorch.org/docs/master/generated/torch.nn.Module.html#torch.nn.Module.state_dict for details.")
+        else:
+            # TODO: remove all above and execute the following block directly after deprecation removal
```
will fix

lkct commented on torch/nn/modules/module.py:
```diff
@@ -1309,19 +1372,43 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
         """
-        if destination is None:
+
+        # Fixes https://github.com/pytorch/pytorch/issues/72778
```
see comment on kw-only

lkct committed:
doc for state_dict

lkct committed:
`Module.state_dict` implementation becomes private, behaviour unchanged

lkct committed:
fix #72778
Deprecate `destination` from `Module.state_dict`.
Deprecate positional args from `Module.state_dict`.

albanD reviewed (commented):
Thanks for the update.
Just small details to reduce noise in the doc and this will be good to go!

albanD reviewed (approved):
Thanks for the update!

Balandat commented on torch/nn/modules/module.py:
```diff
@@ -1278,8 +1278,34 @@ def _save_to_state_dict(self, destination, prefix, keep_vars):
+        self._save_to_state_dict(destination, prefix, keep_vars)
+        for name, module in self._modules.items():
+            if module is not None:
+                module.state_dict(destination, prefix + name + '.', keep_vars=keep_vars)
```
@lkct , @albanD , the way this is implemented here means that whenever a module has a sub-module, calling `state_dict()` (a perfectly fine and non-deprecated thing to do) will raise the deprecation warning. 

Minimal repro: 
```
from torch.nn import Module

m = Module()
m.state_dict()  # this is fine

m.submodule = Module()
m.state_dict()  # this raises the DeprecationWarning (falsely)
```

This should result in this DeprecationWarning being raised left and right in everyone's code who has nested modules (which is pretty much everyone). This is already happening to us. 

cc @dme65, @saitcakmak
